pandas/core/missing.py

```diff
@@ -23,7 +23,9 @@
     from pandas import Index
 
 
-def mask_missing(arr: ArrayLike, values_to_mask) -> np.ndarray:
+def mask_missing(
+    arr: ArrayLike, values_to_mask: Union[List, Tuple, Scalar]
```

Comment:
`values_to_mask` is passed to `infer_dtype_from_array`

so the type of `values_to_mask` should be the same as the `arr` argument of `infer_dtype_from_array`

the `arr` argument of `infer_dtype_from_array` is not yet typed. (sometimes better to type the lower level functions first, but while np.ndarray resolves to `Any`, mypy won't report any inconsistencies)

looking quickly through the body of infer_dtype_from_array, Series and EA are handled, so passing an array-like to `values_to_mask` should be fine.

To handle Series, could either add to Union, or could maybe (if no perf implications) also update `infer_dtype_from_array` to handle Index as well, and use `AnyArrayLike` instead.

Reply:
Adding the Index check to `infer_dtype_from_array`
```
diff --git a/pandas/core/dtypes/cast.py b/pandas/core/dtypes/cast.py
index 165e63e23d..676ab572a0 100644
--- a/pandas/core/dtypes/cast.py
+++ b/pandas/core/dtypes/cast.py
@@ -83,6 +83,7 @@ from pandas.core.dtypes.generic import (
     ABCDatetimeArray,
     ABCDatetimeIndex,
     ABCExtensionArray,
+    ABCIndexClass,
     ABCPeriodArray,
     ABCPeriodIndex,
     ABCSeries,
@@ -877,7 +878,7 @@ def infer_dtype_from_array(
     if pandas_dtype and is_extension_array_dtype(arr):
         return arr.dtype, arr
 
-    elif isinstance(arr, ABCSeries):
+    elif isinstance(arr, ABCSeries) or isinstance(arr, ABCIndexClass):
         return arr.dtype, np.asarray(arr)
 
     # don't force numpy coerce with nan's
```
seems to degrade some of the benchmarks that hit this
```
asv continuous -f 1.1 upstream/master HEAD -b ^algorithms -b ^replace

       before           after         ratio
     [d0db0098]       [9b685722]
     <master~1>       <master>  
+     3.41±0.04ms       5.40±0.7ms     1.59  algorithms.Factorize.time_factorize(False, False, 'boolean')
+     4.50±0.06ms       6.98±0.9ms     1.55  algorithms.Factorize.time_factorize(False, True, 'boolean')
+     3.27±0.01μs       3.89±0.5μs     1.19  algorithms.Duplicated.time_duplicated(True, 'first', 'float')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```
so I guess a `Union` with `Series` is the way to go unless there's a more efficient patch for `infer_dtype_from_array`

Reply:
Thanks for looking into this. Looking again, I see that Index is already handled and falls through the the final return after `np.asarray(arr)`. so could probably use `Union[AnyArrayLike, Sequence[Any], Scalar]` after all.

Reply:
Right - done